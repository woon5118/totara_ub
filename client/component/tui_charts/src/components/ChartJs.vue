<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Aleksandr Baishev <aleksandr.baishev@totaralearning.com>
  @module tui_charts
-->

<template>
  <div>
    <div
      v-if="header && header.trim() !== ''"
      class="tui-chartJs__header"
      :title="header"
    >
      {{ header }}
    </div>
    <canvas
      ref="canvas"
      v-bind="canvasAttributes"
      :aria-label="label"
      role="img"
      @click="clicked"
      @mousemove="hovered"
    >
      <slot :props="{ type, data, options }"><p v-text="label"/></slot>
    </canvas>
  </div>
</template>

<script>
import Chart from 'chart.js';
import theme from 'tui/theme';
import 'chartjs-plugin-doughnutlabel';

const hasOwnProperty = Object.prototype.hasOwnProperty;

Chart.defaults.global.defaultColor = theme.getVar('color-chart-background-1');
Chart.defaults.global.defaultFontSize = parseInt(
  theme.getVar('font-size-chart-default')
);

let defaultColors = [
  {
    // 01
    baseColor: theme.getVar('color-chart-background-1'),
    transparentColor: theme.getVar('color-chart-transparent-1'),
  },
  {
    // 02
    baseColor: theme.getVar('color-chart-background-2'),
    transparentColor: theme.getVar('color-chart-transparent-2'),
  },
  {
    // 03
    baseColor: theme.getVar('color-chart-background-3'),
    transparentColor: theme.getVar('color-chart-transparent-3'),
  },
  {
    // 04
    baseColor: theme.getVar('color-chart-background-4'),
    transparentColor: theme.getVar('color-chart-transparent-4'),
  },
  {
    // 05
    baseColor: theme.getVar('color-chart-background-5'),
    transparentColor: theme.getVar('color-chart-transparent-5'),
  },
  {
    // 06
    baseColor: theme.getVar('color-chart-background-6'),
    transparentColor: theme.getVar('color-chart-transparent-6'),
  },
  {
    // 07
    baseColor: theme.getVar('color-chart-background-7'),
    transparentColor: theme.getVar('color-chart-transparent-7'),
  },
];

const pickColor = function(index) {
  return defaultColors[index % defaultColors.length];
};

const applyUndefinedProperties = function(from, to) {
  for (let def in from) {
    if (typeof to[def] === 'undefined' && hasOwnProperty.call(from, def)) {
      to[def] = from[def];
    }
  }
};

// Since ChartJs doesn't support default colors, we have to implement this functionality manually
// as advised by ChartJs devs https://github.com/chartjs/Chart.js/issues/1618
Chart.pluginService.register({
  beforeUpdate(chart) {
    switch (chart.config.type) {
      case 'bar': {
        // We need to apply colors differently depending on whether the bars are stacked...
        chart.data.datasets.forEach((dataset, index) => {
          if (
            typeof chart.options.scales.xAxes[0].stacked === 'boolean' &&
            chart.options.scales.xAxes[0].stacked &&
            typeof chart.options.scales.yAxes[0].stacked === 'boolean' &&
            chart.options.scales.yAxes[0].stacked
          ) {
            const baseColor = pickColor(index).baseColor;
            let defaults = {
              backgroundColor: baseColor,
            };

            applyUndefinedProperties(defaults, dataset);
          } else {
            if (typeof dataset.backgroundColor !== 'undefined') {
              return;
            }

            let backgroundColors = [];

            dataset.data.forEach((value, index) => {
              backgroundColors.push(pickColor(index).baseColor);
            });

            dataset.backgroundColor = backgroundColors;
          }
        });
        break;
      }

      case 'radar': {
        chart.data.datasets.forEach((dataset, index) => {
          const baseColor = pickColor(index).baseColor;
          let transparentColor = pickColor(index).transparentColor;

          if (typeof transparentColor === 'undefined') {
            transparentColor = baseColor;
          }

          let defaults = {
            borderColor: baseColor,
            backgroundColor: transparentColor,
            pointBackgroundColor: baseColor,
            pointBorderColor: baseColor,
            fill: true,
          };

          applyUndefinedProperties(defaults, dataset);
        });
        break;
      }

      // Fallthrough intended
      case 'line':
      case 'scatter':
      case 'bubble': {
        chart.data.datasets.forEach((dataset, index) => {
          const baseColor = pickColor(index).baseColor;
          let transparentColor = pickColor(index).transparentColor;

          if (typeof transparentColor === 'undefined') {
            transparentColor = baseColor;
          }

          let defaults = {
            borderColor: baseColor,
            backgroundColor: transparentColor,
            pointBackgroundColor: baseColor,
            pointBorderColor: baseColor,
            fill: false,
          };

          applyUndefinedProperties(defaults, dataset);
        });

        break;
      }

      // Fallthrough intended
      case 'doughnut':
      case 'pie':
      case 'polarArea': {
        chart.data.datasets.forEach(dataset => {
          if (!dataset.backgroundColor) {
            dataset.backgroundColor = [];
            dataset.data.forEach((val, index) => {
              dataset.backgroundColor.push(pickColor(index).baseColor);
            });
          }
        });
      }
    }
  },
});

export default {
  name: 'ChartJs',

  props: {
    header: {
      type: String,
      default: '',
    },
    ariaLabel: {
      type: String,
      default: '',
    },
    type: {
      required: true,
      type: String,
      validator: value =>
        [
          'line',
          'bar',
          'radar',
          'pie',
          'doughnut',
          'polarArea',
          'bubble',
          'scatter',
        ].indexOf(value) !== -1,
    },
    data: {
      required: true,
      type: Object,
    },
    options: {
      type: Object,
      default: function() {
        return {};
      },
    },
    canvasAttributes: {
      type: Object,
      default() {
        return {};
      },
    },
  },

  computed: {
    config() {
      return {
        type: this.type,
        data: this.data,
        options: this.options,
      };
    },

    label() {
      return this.ariaLabel.trim() !== '' ? this.ariaLabel : this.header;
    },
  },

  watch: {
    config: {
      deep: true,
      handler() {
        this.renderChart();
      },
    },
    type() {
      this.renderChart();
    },
  },

  mounted() {
    this.renderChart();
  },

  beforeDestroy() {
    this.chart.destroy();
  },

  methods: {
    renderChart() {
      const root = this.$refs.canvas;

      if (this.chart) {
        this.chart.destroy();
      }

      this.chart = new Chart(root, this.config);
    },

    clicked(e) {
      this.$emit(
        'click',
        e,
        this.$refs.canvas.getContext('2d'),
        this.chart,
        Chart
      );
    },

    hovered(e) {
      this.$emit(
        'hover',
        e,
        this.$refs.canvas.getContext('2d'),
        this.chart,
        Chart
      );
    },
  },
};
</script>

<style lang="scss">
:root {
  // Chart default font size
  --font-size-chart-default: 14; // It has to be without units
  --font-size-chart-doughnut-label: 16; // It has to be without units
  --font-size-chart-doughnut-percentage-label: 30; // It has to be without units
}

.tui-chartJs__header {
  @include tui-font-heading-small();
  padding-bottom: var(--gap-7);
  overflow: hidden;
  white-space: nowrap;
  text-align: center;
  text-overflow: ellipsis;
}
</style>
