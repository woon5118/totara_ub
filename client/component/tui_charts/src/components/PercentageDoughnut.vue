<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Aleksandr Baishev <aleksandr.baishev@totaralearning.com>
  @module tui_charts
-->

<template>
  <ChartJs
    type="doughnut"
    :data="data"
    :options="options"
    :header="header"
    :canvas-attributes="canvasAttributes"
    :aria-label="fallback"
  />
</template>

<script>
import ChartJs from 'tui_charts/components/ChartJs';
import theme from 'tui/theme';

let defaultColor = theme.getVar('color-chart-background-1');
let defaultLabelFontSize = parseInt(
  theme.getVar('font-size-chart-doughnut-label')
);
let defaultPercentageFontSize = parseInt(
  theme.getVar('font-size-chart-doughnut-percentage-label')
);
let defaultLabelColor = theme.getVar('color-chart-doughnut-label');
let defaultBackgroundColor = theme.getVar('color-neutral-4');

export default {
  components: { ChartJs },
  props: {
    header: {
      type: String,
    },
    percentage: {
      type: Number,
      required: true,
    },
    label: {
      type: String,
      default: '',
    },
    showPercentage: {
      type: Boolean,
      default: true,
    },
    color: {
      type: String,
    },
    labelColor: {
      type: String,
    },
    labelFontSize: {
      type: Number,
    },
    labelFontWeight: {
      type: [String, Number],
      default: undefined,
    },
    percentageFontSize: {
      type: Number,
    },
    backgroundColor: {
      type: String,
    },
    cutout: {
      type: Number,
      default: 85,
    },
    square: {
      type: Boolean,
      default: false,
    },
  },

  computed: {
    data() {
      let colors = [
        this.color ? this.color : defaultColor,
        this.backgroundColor ? this.backgroundColor : defaultBackgroundColor,
      ];

      let data = [this.percentage, 100 - this.percentage];

      if (this.percentage === 100) {
        colors = [this.color ? this.color : defaultColor];
        data = [this.percentage];
      }

      return {
        datasets: [
          {
            backgroundColor: colors,
            hoverBackgroundColor: colors,
            hoverBorderColor: colors,
            borderWidth: 2,
            borderColor: colors,
            data: data,
          },
        ],
      };
    },

    options() {
      let labels = [];

      if (this.showPercentage) {
        labels.push({
          text: this.percentage + '%',
          font: {
            size: this.percentageFontSize
              ? this.percentageFontSize
              : defaultPercentageFontSize,
            weight: 'bold',
          },
          color: this.labelColor ? this.labelColor : defaultLabelColor,
        });
      }

      if (this.label.trim() !== '') {
        labels.push({
          text: this.label,
          font: {
            size: this.labelFontSize
              ? this.labelFontSize
              : defaultLabelFontSize,
            weight: this.labelFontWeight,
          },
          color: this.labelColor ? this.labelColor : defaultLabelColor,
        });
      }

      let layout = this.square
        ? {
            padding: {
              left: 0,
              right: 0,
              top: 0,
              bottom: 0,
            },
          }
        : {};

      return {
        cutoutPercentage: this.cutout,
        tooltips: {
          enabled: false,
        },
        plugins: {
          doughnutlabel: {
            labels,
          },
        },
        legend: {
          display: false,
        },
        layout,
      };
    },

    canvasAttributes() {
      return this.square
        ? {
            width: '100%',
            height: '100%',
          }
        : {};
    },

    fallback() {
      return (
        (this.label.trim() !== '' ? `${this.label} ` : '') +
        this.percentage +
        '%'
      );
    },
  },
};
</script>
