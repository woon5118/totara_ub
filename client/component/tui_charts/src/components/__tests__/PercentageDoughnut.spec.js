/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Aleksandr Baishev <aleksandr.baishev@totaralearning.com>
 * @module tui
 */

import { mount } from '@vue/test-utils';
import PercentageDoughnut from 'tui_charts/components/PercentageDoughnut';
import ChartJs from 'tui_charts/components/ChartJs';
import { axe, toHaveNoViolations } from 'jest-axe';
expect.extend(toHaveNoViolations);

let wrapper;

// We are mocking theme to override css variables
jest.mock('tui/theme', () => {
  return {
    getVar(name) {
      let vars = {
        'color-chart-background-1': '#3869B1',
        'font-size-chart-doughnut-label': '24',
        'font-size-chart-doughnut-percentage-label': '40',
        'color-chart-doughnut-label': '#000',
        'color-neutral-4': '#e6e4e4',
      };
      return vars[name];
    },
  };
});

describe('PercentageDoughnut', () => {
  beforeAll(() => {
    wrapper = mount(PercentageDoughnut, {
      propsData: {
        percentage: 75,
        header: 'I am a big percentage doughnut',
      },
    });
  });

  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('initializes correct defaults for charts component', () => {
    let data = {
      datasets: [
        {
          data: [75, 25],
          borderWidth: 2,
          borderColor: ['#3869B1', '#e6e4e4'],
          backgroundColor: ['#3869B1', '#e6e4e4'],
          hoverBackgroundColor: ['#3869B1', '#e6e4e4'],
          hoverBorderColor: ['#3869B1', '#e6e4e4'],
        },
      ],
    };

    let options = {
      cutoutPercentage: 85,
      tooltips: {
        enabled: false,
      },
      plugins: {
        doughnutlabel: {
          labels: [
            {
              text: '75%',
              font: {
                size: 40,
                weight: 'bold',
              },
              color: '#000',
            },
          ],
        },
      },
      legend: {
        display: false,
      },
    };

    expect(wrapper.find(ChartJs).props('options')).toMatchObject(options);
    expect(wrapper.find(ChartJs).props('data')).toMatchObject(data);
    expect(wrapper.find(ChartJs).props('header')).toMatch(
      'I am a big percentage doughnut'
    );
    expect(wrapper.find(ChartJs).props('ariaLabel')).toMatch('75%');
  });

  it('passes correct properties to the chart component', () => {
    let doughnut = mount(PercentageDoughnut, {
      propsData: {
        percentage: 57,
        header: 'I am a big percentage doughnut!!',
        square: true,
        color: 'yellow',
        backgroundColor: '#123456',
        labelColor: '#654321',
        label: 'I am a label',
        labelFontSize: 16,
        percentageFontSize: 32,
        cutout: 80,
      },
    });

    let data = {
      datasets: [
        {
          data: [57, 43],
          borderWidth: 2,
          borderColor: ['yellow', '#123456'],
          backgroundColor: ['yellow', '#123456'],
          hoverBackgroundColor: ['yellow', '#123456'],
          hoverBorderColor: ['yellow', '#123456'],
        },
      ],
    };

    let options = {
      cutoutPercentage: 80,
      tooltips: {
        enabled: false,
      },
      plugins: {
        doughnutlabel: {
          labels: [
            {
              text: '57%',
              font: {
                size: 32,
                weight: 'bold',
              },
              color: '#654321',
            },
            {
              text: 'I am a label',
              font: {
                size: 16,
              },
              color: '#654321',
            },
          ],
        },
      },
      legend: {
        display: false,
      },
      layout: {
        padding: {
          left: 0,
          right: 0,
          top: 0,
          bottom: 0,
        },
      },
    };

    expect(doughnut.find(ChartJs).props('canvasAttributes')).toMatchObject({
      width: '100%',
      height: '100%',
    });
    expect(doughnut.find(ChartJs).props('options')).toMatchObject(options);
    expect(doughnut.find(ChartJs).props('data')).toMatchObject(data);
    expect(doughnut.find(ChartJs).props('header')).toMatch(
      'I am a big percentage doughnut!!'
    );
    expect(doughnut.find(ChartJs).props('ariaLabel')).toMatch('57%');
  });

  it('does not include second value in the dataset if 100% passed', () => {
    let doughnut = mount(PercentageDoughnut, {
      propsData: {
        percentage: 100,
        header: 'I am a big percentage doughnut!!',
      },
    });
    let data = {
      datasets: [
        {
          data: [100],
          borderWidth: 2,
          borderColor: ['#3869B1'],
          backgroundColor: ['#3869B1'],
          hoverBackgroundColor: ['#3869B1'],
          hoverBorderColor: ['#3869B1'],
        },
      ],
    };

    let options = {
      cutoutPercentage: 85,
      tooltips: {
        enabled: false,
      },
      plugins: {
        doughnutlabel: {
          labels: [
            {
              text: '100%',
              font: {
                size: 40,
                weight: 'bold',
              },
              color: '#000',
            },
          ],
        },
      },
      legend: {
        display: false,
      },
    };

    expect(doughnut.find(ChartJs).props('options')).toMatchObject(options);
    expect(doughnut.find(ChartJs).props('data')).toMatchObject(data);
    expect(doughnut.find(ChartJs).props('header')).toMatch(
      'I am a big percentage doughnut'
    );
    expect(doughnut.find(ChartJs).props('ariaLabel')).toMatch('100%');
  });

  it('should not have any accessibility violations', async () => {
    const results = await axe(wrapper.element, {
      rules: {
        region: { enabled: false },
      },
    });
    expect(results).toHaveNoViolations();
  });
});
