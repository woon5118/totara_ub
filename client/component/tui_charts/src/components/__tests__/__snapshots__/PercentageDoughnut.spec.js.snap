// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PercentageDoughnut renders correctly 1`] = `
<div>
  <div
    class="tui-chartJs__header"
    title="I am a big percentage doughnut"
  >
    
    I am a big percentage doughnut
  
  </div>
   
  <canvas
    aria-label="75%"
    class="chartjs-render-monitor"
    height="0"
    role="img"
    style="display: block;"
    width="0"
  >
    <p>
      75%
    </p>
  </canvas>
</div>
`;